### ISSUE TITLE
Fails to process Alias references

### ISSUE THREAD

**Issue by vicp-iq** [mikeclayton/MofParser#7]
The following would fail:

```
instance of TYPEA as $Alias0000006E
{
    ServerURL = "https://URL";
};

instance of TYPEB as $Alias00000070
{
    Reference = {$Alias0000006E};
};
```

**Comment by mikeclayton** [comment]
Hi, thanks for the report.

I think the issue is more to do with Array parsing rather than Aliases. It probably doesn't help, but if you just use

```
Reference = $Alias0000006E;
```

the parser will quite happily insert a ReferenceTypeValueAst into the tree (see additional parser unit tests that I've just added).

The problem is in PropertyValueAst.Parse(). It does't really handle arrays very well, and currently assumes ComplexValues are *always* arrays, References and Enums are *never* arrays, and Literals *might* be an array.

I've committed a refinement to the comments to make it clearer what spec actually says, and I'm happy to take a look at the code tp try to fix it, but it might be a few days before I can get round to it.

**Comment by vicp-iq** [comment]
That sounds great,  I'll be able to work on PRs on some Friday afternoons, but not all of them.  I hope to use this tool to parse the ConfigurationStatus reports that DSC outputs after completion.

**Comment by mikeclayton** [comment]
I've made some major changes to backtracking to clear the path for fixing this, but it's not quite there yet - see the "feature/ParserState" branch for latest (broken) code.

The Parser now has a stack of ParseStates that hold the position in the token stream, and a bunch of the the Ast classes have a TryParse method that Pushes a new state onto the stack and Pops it off again if they fail to read from the stream, which is easier than remembering how many tokens to backtrack.

I'll do some more work to get this compiling and running over the next couple of days.

**Comment by mikeclayton** [comment]
I've stabilised the "feature/ParserState" branch now, and it passes some Parser tests I added for this issue. Could you try it and let me know if it looks ok with your mof files, and I'll merge into master.

**Comment by vicp-iq** [comment]
Sure, I'll give that a shot today.

**Comment by vicp-iq** [comment]
Have you reviewed my PRs yet?  Seems like you made some pretty significant refactoring changes that introduced some challenging merge conflicts.

**Comment by vicp-iq** [comment]
This line is defective: https://github.com/mikeclayton/MofParser/blob/feature/ParserState/src/Kingsland.MofParser/Ast/ReferenceValueArrayAst.cs#L94.  

It took a while to track down because if you try to view results in a debug window, it crashes the debug session without any output.  By adding the following line `var stuff = result.ToString();` it is revealed that something something is causing the method to throw a StackOverflow Exception.

**Comment by vicp-iq** [comment]
My Pull Requests have been sitting unresolved for over 2 years, and I'm now prioritizing the work that uses this library.  I'm willing to contribute back to this repo, but only if my PRs are going to get reviewed and processed in a timely manner.  If you say nothing, then I'll just fork it and move on, and if you tell me you'd like me to continue creating PR's then I'll do that

**Comment by vicp-iq** [comment]
I'm prioritizing this work again... Just wanted to get some clarity about expectations.  Are you waiting on me right now?

**Comment by mikeclayton** [comment]
This should be fixed in the latest master branch - the MOF 3.0.1 spec has some changes and corrections in the documentation that I've implemented, and I've added support for property values that contain arrays.
